File: LeetCode/problems/p0021/python/p021.py
#[COMPLETED]

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def print_ln(self):
        print(self.val, end="")
        if self.next is not None:
            print(" -> ", end="")
            self.next.print_ln()
        else:
            print()

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if l1 is None:
            return l2
        elif l2 is None:
            return l1

        head = None
        if l1.val < l2.val:
            head = l1
            l1 = l1.next
        else:
            head = l2
            l2 = l2.next

        trav = head

        while l1 is not None and l2 is not None:
            if l1.val < l2.val:
                trav.next = l1
                l1 = l1.next
            else:
                trav.next = l2
                l2 = l2.next
            trav = trav.next

        if l1 is not None:
            trav.next = l1
        elif l2 is not None:
            trav.next = l2

        return head


sol = Solution()
l1 = ListNode(1)
l1.next = ListNode(2)
l1.next.next = ListNode(4)
l1.next.next.next = ListNode(10)

l2 = ListNode(1)
l2.next = ListNode(3)
l2.next.next = ListNode(4)

ret = sol.mergeTwoLists(l1, l2)
if ret is not None:
    ret.print_ln()
else:
    print("None")
print("Expected 1->1->2->3->4->4")

File: leet-code/src/problems/p0021/java/src/p021/java/P021.java
//[COMPLETED]

package p021;

public class P021 {

    public static void main(String[] args) {
        new P021().solve();
    }
    
    public void solve() {
        ListNode l1 = new ListNode(1);
        l1.next = new ListNode(2);
        l1.next.next = new ListNode(4);
        
        ListNode l2 = new ListNode(1);
        l2.next = new ListNode(3);
        l2.next.next = new ListNode(4);
        
        System.out.println(l1);
        System.out.println(l2);
        
        
        Solution s = new Solution();
        System.out.println(s.mergeTwoLists(l1, l2).toString());
        
        
    }
    
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
        
        @Override public String toString() {
            if (this.next == null)
                return "" + val;
            return "" + val + " -> " + this.next.toString();
        }
    }   
 
 
private class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null)  return null;
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        
        ListNode result;
        
        if (l1.val < l2.val) {
            result = new ListNode(l1.val);
            l1 = l1.next;
        }
        else {
            result = new ListNode(l2.val);
            l2 = l2.next;
        }
        
        ListNode resultIter = result;
        
        while (l1 != null || l2 != null) {
            if (l1 == null) {
                resultIter.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            else if (l2 == null || l1.val < l2.val) {
                resultIter.next = new ListNode(l1.val);
                l1 = l1.next;
            }
            else {
                resultIter.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            resultIter = resultIter.next;  
        }
        
        return result;
    }
}

}
